MediaCodecVideoRenderer error

### Describe the bug

This is something I have experienced multiple times now. After uploading some new files to my Jellyfin server, I get this error when trying to play them. I am pretty certain it is the same bug that is also described in this thread:
https://github.com/google/ExoPlayer/issues/8509
It is important to note that I can still play files I uploaded some longer time ago, this only affects the newly uploaded files. This leads me to believe it is the same bug described on this post of the Jellyfin forum:
https://forum.jellyfin.org/t-solved-all-new-videos-won-t-play-anymore
I first thought I waa experiencing the same issue as the user in this thread is describing:
https://github.com/jellyfin/jellyfin-android/issues/1014
however he is getting an unsupported format error, the error I'm getting is different.
![Screenshot_20241016_031412_Jellyfin](https://github.com/user-attachments/assets/b4dad368-ecb7-4762-b322-8c8539dc8b0c)


### Logs

_No response_

### Application version

2.6.2

### Where did you install the app from?

Sideloaded APK (proprietary build)

### Device information

Samsung Galaxy A23

### Android version

Android 14

### Jellyfin server version

10.9.11

### Which video player implementations does this bug apply to?

- [ ] Web player (default)
- [X] Integrated player (ExoPlayer)
- [ ] External player (VLC, mpv, MX Player)

I'm having this issue on my S24 Ultra with h264 10 bit video on exoplayer. MPV works fine.
I ran jellyfin-android through android studio with ADB debugging, and it looks like the problem for me is that the AVC decoder doesn't support h264 10 bit. This would seem to be an exoplayer or samsung or qualcomm issue. However, maybe jellyfin can transcode to h264 8 bit, or another viable format, if it detects that the decode failed.
```
2024-12-05 00:03:26.443 18685-27784 CCodec                  org.jellyfin.mobile.debug            I  [c2.qti.avc.decoder] state->set(RUNNING)
2024-12-05 00:03:26.443 18685-27784 CCodecBufferChannel     org.jellyfin.mobile.debug            I  [c2.qti.avc.decoder#779] 10 initial input buffers available
2024-12-05 00:03:26.444 18685-27780 MediaCodecInfo          org.jellyfin.mobile.debug            D  NoSupport [codec.profileLevel, avc1.6E0034, video/avc] [c2.qti.avc.decoder, video/avc] [e3q, SM-S928U, samsung, 34]
2024-12-05 00:03:26.444 18685-27780 MediaCodecRenderer      org.jellyfin.mobile.debug            W  Format exceeds selected codec's capabilities [id=1, mimeType=video/avc, codecs=avc1.6E0034, res=1920x1080, language=en, selectionFlags=[default], c2.qti.avc.decoder]
2024-12-05 00:03:26.445 18685-27783 MediaCodec              org.jellyfin.mobile.debug            I  setCodecState state(0), called in 6
2024-12-05 00:03:26.451 18685-19808 CCodec                  org.jellyfin.mobile.debug            D  Component "c2.qti.avc.decoder" returned error: 0xffffffff
2024-12-05 00:03:26.452 18685-27783 MediaCodec              org.jellyfin.mobile.debug            E  Codec reported err 0x80000000/UNKNOWN_ERROR, actionCode 0, while in state 6/STARTED
2024-12-05 00:03:26.452 18685-27783 MediaCodec              org.jellyfin.mobile.debug            D  flushMediametrics
2024-12-05 00:03:26.452 18685-27783 SurfaceUtils            org.jellyfin.mobile.debug            D  disconnecting from surface 0xb400007bdf87fbf0, reason disconnectFromSurface
2024-12-05 00:03:26.452 18685-27783 MediaCodec              org.jellyfin.mobile.debug            E  queueInputBuffer() is valid only at Executing states; currently at Released state
2024-12-05 00:03:26.452 18685-27783 MediaCodec              org.jellyfin.mobile.debug            E  queueInputBuffer() is valid only at Executing states; currently at Released state
2024-12-05 00:03:26.453 18685-27783 MediaCodec              org.jellyfin.mobile.debug            E  queueInputBuffer() is valid only at Executing states; currently at Released state
2024-12-05 00:03:26.453 18685-27783 MediaCodec              org.jellyfin.mobile.debug            E  queueInputBuffer() is valid only at Executing states; currently at Released state
2024-12-05 00:03:26.453 18685-27783 MediaCodec              org.jellyfin.mobile.debug            E  queueInputBuffer() is valid only at Executing states; currently at Released state
2024-12-05 00:03:26.453 18685-27783 MediaCodec              org.jellyfin.mobile.debug            E  queueInputBuffer() is valid only at Executing states; currently at Released state
2024-12-05 00:03:26.453 18685-27783 MediaCodec              org.jellyfin.mobile.debug            E  queueInputBuffer() is valid only at Executing states; currently at Released state
2024-12-05 00:03:26.453 18685-27783 MediaCodec              org.jellyfin.mobile.debug            E  queueInputBuffer() is valid only at Executing states; currently at Released state
2024-12-05 00:03:26.459 18685-27780 MediaCodecVideoRenderer org.jellyfin.mobile.debug            E  Video codec error
                                                                                                      java.lang.IllegalStateException: queueInputBuffer() is valid only at Executing states; currently at Released state
                                                                                                      
                                                                                                          at android.media.MediaCodec.native_queueInputBuffer(Native Method)
                                                                                                          at android.media.MediaCodec.queueInputBuffer(MediaCodec.java:2831)
                                                                                                          at com.google.android.exoplayer2.mediacodec.AsynchronousMediaCodecBufferEnqueuer.doQueueInputBuffer(AsynchronousMediaCodecBufferEnqueuer.java:233)
                                                                                                          at com.google.android.exoplayer2.mediacodec.AsynchronousMediaCodecBufferEnqueuer.doHandleMessage(AsynchronousMediaCodecBufferEnqueuer.java:206)
                                                                                                          at com.google.android.exoplayer2.mediacodec.AsynchronousMediaCodecBufferEnqueuer.access$000(AsynchronousMediaCodecBufferEnqueuer.java:53)
                                                                                                          at com.google.android.exoplayer2.mediacodec.AsynchronousMediaCodecBufferEnqueuer$1.handleMessage(AsynchronousMediaCodecBufferEnqueuer.java:102)
                                                                                                          at android.os.Handler.dispatchMessage(Handler.java:106)
                                                                                                          at android.os.Looper.loopOnce(Looper.java:230)
                                                                                                          at android.os.Looper.loop(Looper.java:319)
                                                                                                          at android.os.HandlerThread.run(HandlerThread.java:67)
2024-12-05 00:03:26.459 18685-27780 MediaCodecInfo          org.jellyfin.mobile.debug            D  NoSupport [codec.profileLevel, avc1.6E0034, video/avc] [c2.qti.avc.decoder, video/avc] [e3q, SM-S928U, samsung, 34]
2024-12-05 00:03:26.459 18685-27780 MediaCodecInfo          org.jellyfin.mobile.debug            D  NoSupport [codec.profileLevel, avc1.6E0034, video/avc] [c2.qti.avc.decoder.low_latency, video/avc] [e3q, SM-S928U, samsung, 34]
2024-12-05 00:03:26.460 18685-27780 ExoPlayerImplInternal   org.jellyfin.mobile.debug            E  Playback error
                                                                                                      com.google.android.exoplayer2.ExoPlaybackException: MediaCodecVideoRenderer error, index=0, format=Format(1, null, null, video/avc, avc1.6E0034, -1, en, [1920, 1080, -1.0, null], [-1, -1]), format_supported=NO_EXCEEDS_CAPABILITIES
                                                                                                          at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:601)
                                                                                                          at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                                          at android.os.Looper.loopOnce(Looper.java:230)
                                                                                                          at android.os.Looper.loop(Looper.java:319)
                                                                                                          at android.os.HandlerThread.run(HandlerThread.java:67)
                                                                                                      Caused by: com.google.android.exoplayer2.video.MediaCodecVideoDecoderException: Decoder failed: c2.qti.avc.decoder
                                                                                                          at com.google.android.exoplayer2.video.MediaCodecVideoRenderer.createDecoderException(MediaCodecVideoRenderer.java:1864)
                                                                                                          at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:843)
                                                                                                          at com.google.android.exoplayer2.video.MediaCodecVideoRenderer.render(MediaCodecVideoRenderer.java:854)
                                                                                                          at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:1047)
                                                                                                          at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:522)
```

--- a/app/src/main/java/org/jellyfin/mobile/player/deviceprofile/DeviceProfileBuilder.kt
+++ b/app/src/main/java/org/jellyfin/mobile/player/deviceprofile/DeviceProfileBuilder.kt
@@ -4,11 +4,15 @@ import android.media.MediaCodecList
 import org.jellyfin.mobile.app.AppPreferences
 import org.jellyfin.mobile.utils.Constants
 import org.jellyfin.sdk.model.api.CodecProfile
+import org.jellyfin.sdk.model.api.CodecType
 import org.jellyfin.sdk.model.api.ContainerProfile
 import org.jellyfin.sdk.model.api.DeviceProfile
 import org.jellyfin.sdk.model.api.DirectPlayProfile
 import org.jellyfin.sdk.model.api.DlnaProfileType
 import org.jellyfin.sdk.model.api.MediaStreamProtocol
+import org.jellyfin.sdk.model.api.ProfileCondition
+import org.jellyfin.sdk.model.api.ProfileConditionType
+import org.jellyfin.sdk.model.api.ProfileConditionValue
 import org.jellyfin.sdk.model.api.SubtitleDeliveryMethod
 import org.jellyfin.sdk.model.api.SubtitleProfile
 import org.jellyfin.sdk.model.api.TranscodingProfile
@@ -18,6 +22,7 @@ class DeviceProfileBuilder(
 ) {
     private val supportedVideoCodecs: Array<Array<String>>
     private val supportedAudioCodecs: Array<Array<String>>
+    private val videoCodecsProfiles: Map<String, Set<String>>
 
     private val transcodingProfiles: List<TranscodingProfile>
 
@@ -66,6 +71,7 @@ class DeviceProfileBuilder(
                 audioCodecs.containsKey(codec) || codec in FORCED_AUDIO_CODECS
             }.toTypedArray()
         }
+        videoCodecsProfiles = videoCodecs.entries.associate { (k, v) -> k to v.profiles }
 
         transcodingProfiles = listOf(
             TranscodingProfile(
@@ -109,11 +115,14 @@ class DeviceProfileBuilder(
                 directPlayProfiles.add(
                     DirectPlayProfile(
                         type = DlnaProfileType.VIDEO,
-                        container = SUPPORTED_CONTAINER_FORMATS[i],
+                        container = container,
                         videoCodec = supportedVideoCodecs[i].joinToString(","),
                         audioCodec = supportedAudioCodecs[i].joinToString(","),
                     ),
                 )
+                for (videoCodec in supportedVideoCodecs[i]) {
+                    generateCodecProfile(container, videoCodec)?.let(codecProfiles::add)
+                }
             }
             if (supportedAudioCodecs[i].isNotEmpty()) {
                 containerProfiles.add(
@@ -146,6 +155,31 @@ class DeviceProfileBuilder(
             maxStreamingBitrate = MAX_STREAMING_BITRATE,
             maxStaticBitrate = MAX_STATIC_BITRATE,
             musicStreamingTranscodingBitrate = MAX_MUSIC_TRANSCODING_BITRATE,
+        )
+    }
+
+    private fun generateCodecProfile(
+        container: String,
+        videoCodec: String,
+    ): CodecProfile? {
+        val profilesSet = videoCodecsProfiles[videoCodec]
+        if (profilesSet?.isNotEmpty() != true) {
+            return null
+        }
+
+        return CodecProfile(
+            type = CodecType.VIDEO,
+            container = container,
+            codec = videoCodec,
+            applyConditions = listOf(),
+            conditions = listOf(
+                ProfileCondition(
+                    condition = ProfileConditionType.EQUALS_ANY,
+                    property = ProfileConditionValue.VIDEO_PROFILE,
+                    value = profilesSet.joinToString("|"),
+                    isRequired = false,
+                ),
+            ),
         )
     }
 

